Add filtering functionality to the symbols list

LeonarddeR opened the pull request:
### Link to issue number:
Closes #5761

### Summary of the issue:
Now that emoji's have been introduced in NVDA symbol dictionaries, the list of symbols was getting pretty big. This also introduced a major lag when opening the speech symbols dialog.

### Description of how this pull request fixes the issue:
1. Added a filter edit box to the speech symbols dialog. @feerrenrut: your opinion about the visuals of this edit box are greatly appreciated. Feel free to push directly to this branch if you need to change things visually.
2. Converted the list control to a list control with the wx.LC_VIRTUAL style. I'm really exited about the major performance boost this gives when loading the dialog for the first time, as well as when filtering. It is just really fast now!
3. Added some functionality to gui.nvdaControls.AutoWidthColumnListCtrl to support getting item text from a data storage without the requirement of subclassing the list control. It is now also possible to forcefully send a wx.EVT_LIST_ITEM_FOCUSED to the control

### Testing performed:
Filtering, adding and removing items, closing the dialog and reopening it again quickly after that.

### Known issues with pull request:
* The visuals might not be ideal yet, though guiHelper is really amazing in creating stuf that just looks good.
* We should probably wait for #8006 to be in master and apply scaling support to this dialog as well while at it.
* This does not yet apply to speech dictionaries. I think it should at some point, but that won't be part of this pr.

### Change log entry:
* New features
    + You can now filter symbols in the punctuation/symbol pronunciation dialog, similar to how filtering works in the elements list and input gestures dialog. (#5761)

* Bug fixes
   + Loading the punctuation/symbol pronunciation dialog is now much faster when using symbol dictionaries containing over 1000 entries.

Commit:
Add filtering functionality to the symbols list

leonardder commented:
Wow! I'm coming across like 31000 symbols here, taking like 20 seconds to load with the current Alpha version of NVDA. This pr reduces that to 2 seconds.

Commit:
Removed extra line at end of nvdaControls which was accidentally added

Commit:
Default the filter to an empty string instead of None

Commit:
Updated user guide

dnz3d4c commented:
Thanks.
> 
> 
> Wow! I'm coming across like 31000 symbols here, taking like 20 seconds to load with the current Alpha version of NVDA. This pr reduces that to 2 seconds.

dingpengyu commented:
@Hi leonardder I am the maintainer of NVDA Simplified Chinese. I found that NVDA can't generate Simplified Chinese emoticon dictionary using the scons command.

josephsl commented:
Hi, that’ll be because it isn’t clear at the first glance which XML file is for which Chinese variant. This is something that should be addressed sometime in October. Thanks.

 

From: 星空 <notifications@github.com> 
Sent: Thursday, September 27, 2018 3:58 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [nvaccess/nvda] Add filtering functionality to the symbols list (#8790)

 

@hi <https://github.com/hi>  leonardder I am the maintainer of NVDA Simplified Chinese. I found that NVDA can't generate Simplified Chinese emoticon dictionary using the scons command.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8790#issuecomment-425268399> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkHWIAvP1W5VHrnyPs5yM5apNlItdks5ufVf-gaJpZM4W8gWm> .

dingpengyu commented:
thank josephsl

leonardder commented:
I must have missed the Chinese annotations.

@dingpengyu: Could you please open a new issue for this? I'm not yet known with the various differences in the Chinese language, so it is quite difficult to decided what would be the best assignments. In NVDA, we have the zh-cn, zh-hk and zh-tw locales.

josephsl commented:
Hi, I can assist with this, as I work with Asian languages in a regular basis. When creating this new issue, please at mention me (CC me). Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Thursday, September 27, 2018 10:57 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Comment <comment@noreply.github.com>
Subject: Re: [nvaccess/nvda] Add filtering functionality to the symbols list (#8790)

 

I must have missed the Chinese annotations.

@dingpengyu <https://github.com/dingpengyu> : Could you please open a new issue for this? I'm not yet known with the various differences in the Chinese language, so it is quite difficult to decided what would be the best assignments. In NVDA, we have the zh-cn, zh-hk and zh-tw locales.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8790#issuecomment-425329465> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkLuXnwayHCO0IohnKh3AlhjEipB5ks5ufboXgaJpZM4W8gWm> .

dingpengyu commented:
josephsl@

dingpengyu commented:
Hi, thanks to josephsl's reply, this is the issue address:
NVDA cannot generate Simplified Chinese emoticon dictionary using the scons command · Issue #8799
https://github.com/nvaccess/nvda/issues/8799

leonardder commented:
@feerrenrut: I've removed the blocked label since #8006 is merged. However, I'm not yet sure what to do to make this dialog scale correctly, other than setting resizable to True. The size of the list control and its columns is undefined.

feerrenrut reviewed (changes requested):
When filtering, the list does not update. The entries only update as they gain focus. Visually this is quite confusing.

Because this dialog is a `SettingsDialog` Then there is no need to do anything to get the DPI scaling, except, if you specify a size manually, you must pass it through `self.scaleSize()`. 

I think this dialog could be wider, because the text in the replacement column is often long enough to exceed the width of the column. The replacement field (once populated) often isn't wide enough. There is plenty of space to widen this field, and the dialog itself isn't very wide either.

Some emoji don't seem to be supported by the font. Visually they are just squares. Not a blocking issue right now, since I believe this is already the case. An issue should be created for this.

I can take a look at these points.

leonardder commented:
> When filtering, the list does not update. The entries only update as they gain focus. Visually this is quite confusing.

Ugh, might this be because the list is virtual? Does this also mean that there's only one item visible at a time or something like that?

feerrenrut committed:
Refresh visible items after filter is changed.

Otherwise the items only change when they receive focus. It appears (visually)
as if the list has not changed, until you start to scroll or select items.

Commit:
Merge remote-tracking branch 'origin/master' into filterSymbols

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -2596,30 +2596,76 @@ def wrapWithEventSkip(event):
+		if not filterText:
+			self.filteredSymbols = self.symbols
+		else:
+			#Do case-insensitive matching by lowering both filterText and each symbols's text.
```
Mm, this makes sense, but might be a bit out of scope for this pr. See also https://www.python-course.eu/levenshtein_distance.php . it seems to need a new helper function.

Commit:
Review actions

Commit:
As requested, add the possibility to provide a callable to get the item text at construction time of an AutoWidthColumnListCtrl.
A weak reference to the callable is saved on the control. If we wouldn't, closing the punctuation dialog and reopening it would result in a multi instance error.
The extensionPoints module provides functionality to deal with instance method weakrefs. Updated that code to allow weak references without an unregister callback.

leonardder commented on source/gui/nvdaControls.py:
```diff
@@ -26,6 +29,18 @@ def __init__(self, parent, id=wx.ID_ANY, autoSizeColumnIndex="LAST", pos=wx.Defa
 
+	def OnGetItemText(self, item, column):
+		try:
+			return self.Parent.getItemTextForList(self, item, column)
```
Agreed, though it was not as simple as I thought before.

leonardder commented:
@feerrenrut: Thanks for fixing the list updating issue. Would be great if you can fix the replacement column width as well, since it would really help to do that with a pair of eyes.

feerrenrut committed:
Tidy up list filtering.

feerrenrut commented on source/gui/nvdaControls.py:
```diff
@@ -44,17 +46,23 @@ def __init__(
+				raise TypeError("itemTextCallable should be None or a callable")
+			# We need to store a weak reference to the callable.
+			# If we don't and the callable is a bound instance method of a wx Window,
+			# we might create a circular reference.
```
Another option here is to bind to the window destroy event and clean up.

```
@@ -50,21 +49,27 @@ class AutoWidthColumnListCtrl(wx.ListCtrl, listmix.ListCtrlAutoWidthMixin):
 			if not isinstance(itemTextCallable, Callable):
 				raise TypeError("itemTextCallable should be None or a callable")
 			# We need to store a weak reference to the callable.
 			# If we don't and the callable is a bound instance method of a wx Window,
 			# we might create a circular reference.
-			self._itemTextCallable = getWeakrefForCallable(itemTextCallable)
+			self._itemTextCallable = itemTextCallable
 		else:
-			self._itemTextCallable = None
+			self._itemTextCallable = self._super_itemTextCallable
 		wx.ListCtrl.__init__(self, parent, id=id, pos=pos, size=size, style=style)
 		listmix.ListCtrlAutoWidthMixin.__init__(self)
 		self.setResizeColumn(autoSizeColumnIndex)
+		self.Bind(wx.EVT_WINDOW_DESTROY, source=self, id=self.GetId, handler=self._onDestroy)
+
+	def _onDestroy(self, evt):
+		evt.Skip()
+		self._itemTextCallable = None
+

+	def _super_itemTextCallable(self, item, column):
+		return super(AutoWidthColumnListCtrl, self).OnGetItemText(item, column)

 	def OnGetItemText(self, item, column):
-		if self._itemTextCallable is None or self._itemTextCallable() is None:
-			return super(AutoWidthColumnListCtrl, self).OnGetItemText(item, column)
-		return self._itemTextCallable()(item, column)
+		return self._itemTextCallable(item, column)

 	def sendListItemFocusedEvent(self, index):
 		evt = wx.ListEvent(wx.wxEVT_LIST_ITEM_FOCUSED, self.Id)
```

leonardder replied:
This works as well.

feerrenrut commented:
Yep, I'll take a look at the widths in the dialog tomorrow.

Commit:
Review action, revert changes to extensionPoints

feerrenrut committed:
Fix resizing of the symbol pronunciation dialog.

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into addFilteringPR8790

leonardder commented on source/gui/guiHelper.py:
```diff
@@ -152,12 +152,19 @@ def __init__(self, parent, labelText, wxCtrlClass, **kwargs):
 			@param labelText: The text to associate with a wx control.
 			@type labelText: string
 			@param wxCtrlClass: The class to associate with the label, eg: wx.TextCtrl
+			@type wxCtrlClass: class
```
Is this intentional?

feerrenrut reviewed (approved):
I've looked through this again and I'm happy for it to be merged. @leonardder Are you happy with the latest changes?

leonardder commented on source/gui/guiHelper.py:
```diff
@@ -167,6 +174,21 @@ def control(self):
 	def sizer(self):
 		return self._sizer
 
+class LabeledTextCtrl(LabeledControlHelper):
```
Could you give this a doc string that clarifies why this subclass has been created?

feerrenrut replied:
Yes, I'll do this. Trying to achieve this small variation has made me think carefully about the implementation here. I think we should start to move this towards a more component driven implementation. The initial goal was to try to streamline the GUI code, making it simpler (and less error prone) to do the common things, I think this has been successful. Though now I would like to make this a little more flexible too.

feerrenrut replied:
I decided to remove this class, trying to find a way to implement this variation in a clean way has inspired me to take another approach. For now, this means that two textCtrls that should grow with the dialog will not. I don't think this PR should be blocked on this.

leonardder commented on source/gui/guiHelper.py:
```diff
@@ -167,6 +174,21 @@ def control(self):
+		if(not self.expandWidth):
+			sizer.Add(wxCtrl)
+		else:
+			sizer.Add(wxCtrl, proportion=1)
```
What is this proportion=1? Could a constant help here?

feerrenrut replied:
Proportion is an argument to wx box sizer, it lets you control how much of the sizers orientation a certain element gets. So for a horizontal sizer, it controls the proportion of width allocated to that particular control. Approximately, it seems that the proportion of any unused space is divided amongst the controls that was added to the sizer with a proportion value, based on the total of all the proportion values. I'm not sure that a constant really makes sense. In this particular case, I think any value has the same effect as any other value, though I'm far from certain about this.

LeonarddeR reviewed (commented):
I've got some comments. Also, you could consider adding a convenience wrapper for the new LabeledTextControl to BoxSizerHelper.

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -2638,32 +2638,58 @@ def __init__(self,parent):
+		self.symbolsList = sHelper.addLabeledControl(
+			symbolsText,
+			nvdaControls.AutoWidthColumnListCtrl,
+			autoSizeColumnIndex=2,  # The replacement column is likely to need the most space
```
The interesting thing here is that column with index 2 is level, not replacement. See below.

feerrenrut replied:
Hmm yes, index is really not a good name for this.

dingpengyu commented:
hi
Is there a developer review for this PR? Seeing that it is already approved

feerrenrut committed:
Revert new control

This was a clumsy attempt to be able to support making the textCtrls resize
with the dialog. The real problem is that GuiHelper is not felixible enough
to do it in a simple way. For now, we will just set a reasonable starting
size.

feerrenrut committed:
rename misleading parameter

The parameter was deceptively named "index" where it affected columns
numbers, first column is 1 not 0.

feerrenrut committed:
Simplify selection restoration logic.

Reduce nesting, clarify some comments.

feerrenrut commented:
@leonardder I've made a few changes here, if you are happy with this I think it is ready to be merged.

DataTriny commented:
I didn't mean to make any UI design work on this PR. I'm happy with the code, let's wait for Leonard.

Commit:
Merge remote-tracking branch 'origin/master' into filterSymbols

Commit:
Removing the last item of a filtered list, make sure that we don't try to focus a nonexistent item. Also properly disable the remove button when disabling the controls

leonardder commented:
@feerrenrut: I'm ok with your changes and did a small round of tests.

Found and fixed the following:

1. Add a new symbol, for example called `test`
2. Filter on test
3. Remove the test symbol

What happened is that the list of symbols was empty, yet the old logic tried to select a new symbol. Removing a symbol didn't check for the item count properly. Now it does.

feerrenrut reviewed (approved):
Looks good. I'll merge this in.
